kernel/src/ipc.rs

```diff
@@ -11,33 +11,41 @@ use crate::mem::{AppSlice, Shared};
 use crate::process;
 use crate::returncode::ReturnCode;
 use crate::sched::Kernel;
+use core::mem::MaybeUninit;
 
 /// Syscall number
 pub const DRIVER_NUM: usize = 0x10000;
 
-struct IPCData {
-    shared_memory: [Option<AppSlice<Shared, u8>>; 8],
-    client_callbacks: [Option<Callback>; 8],
+struct IPCData<const NUM_PROCS: usize> {
+    shared_memory: [Option<AppSlice<Shared, u8>>; NUM_PROCS],
+    client_callbacks: [Option<Callback>; NUM_PROCS],
     callback: Option<Callback>,
 }
 
-impl Default for IPCData {
-    fn default() -> IPCData {
-        IPCData {
-            shared_memory: [None, None, None, None, None, None, None, None],
-            client_callbacks: [None, None, None, None, None, None, None, None],
+impl<const NUM_PROCS: usize> Default for IPCData<NUM_PROCS> {
+    fn default() -> IPCData<NUM_PROCS> {
+        let mut ipc_data = IPCData {
+            shared_memory: unsafe { MaybeUninit::zeroed().assume_init() },
+            client_callbacks: unsafe { MaybeUninit::zeroed().assume_init() },
```

Comment:
This really requires `unsafe`?

Reply:
Yes  `assume_init` is unsafe. It is `UB` calling `assume_init` on uninitialized data

Reply:
It just doesn't seem like initializing an array should require an unsafe operation.

Reply:
Can you use `Default::default()` to get the same semantics here without unsafe?

Reply:
Unfortunately, no, that gives:
```
error[E0277]: the trait bound `[core::option::Option<callback::Callback>; _]: core::default::Default` is not satisfied
```

The problem is that `Default` is only implemented for arrays of length <= 32, and as far as I know there is no way for me to specify that the const generic `NUM_PROCS` will always be less than 32.

This issue is known and being worked on (https://github.com/rust-lang/rust/issues/61415)

Reply:
The other solution you might want:

```
        IPCData {
            shared_memory: [None; NUM_PROCS],
            client_callbacks: [None; NUM_PROCS],
            callback: None,
        }
```

Also fails, with the error message `error: array lengths can't depend on generic parameters`. This is also a known issue: https://github.com/rust-lang/rust/issues/68567

Reply:
> It just doesn't seem like initializing an array should require an unsafe operation.

Yepp, agree we need something like https://github.com/rust-lang/rfcs/issues/2626, but using `std::mem::zeroed()` is quite hard to get right in general IMO.

Reply:
@niklasad1 is right, (https://doc.rust-lang.org/core/mem/union.MaybeUninit.html), using `MaybeUninit:zeroed().assume_init()` is the equivalent of the (now deprecated) `mem::zeroed()`.

From reading [here](https://d3m3vilurr.gitbooks.io/the-unsafe-rust-programming-language/uninitialized.html) it seems the correct way to initialize memory in this way is using `ptr::write` to ensure the compiler does not insert any attempts to read the uninitialized values before writing. So I am pushing an update that does that.